src/Agent.Plugins/BuildDrop/BuildDropClient.cs

```diff
@@ -0,0 +1,74 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.BlobStore.Common;
+using Microsoft.VisualStudio.Services.Content.Common.Tracing;
+using Microsoft.VisualStudio.Services.BlobStore.WebApi;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Newtonsoft.Json;
+using Agent.Sdk;
+
+namespace Agent.Plugins.BuildDrop
+{    
+    // A wrapper of BuildDropManager, providing basic functionalities such as uploading and downloading drop artifacts.
+    public class BuildDropClient
+    {
+        // Upload from target path to VSTS BlobStore service through BuildDropManager, then associate it with the build
+        internal async Task UploadDropArtifactAsync(
+            AgentTaskPluginExecutionContext context,
+            VssConnection connection,
+            Guid projectId,
+            int buildId,
+            string name,
+            string source,
+            CancellationToken cancellationToken)
+        {
+            var httpclient = connection.GetClient<DedupStoreHttpClient>();
+            var tracer = new CallbackAppTraceSource(str => context.Output(str), System.Diagnostics.SourceLevels.Information);
+            httpclient.SetTracer(tracer);
+            var client = new DedupStoreClientWithDataport(httpclient, 16 * Environment.ProcessorCount);
```

Comment:
what is 16 * processCount for? parallel dedup calculation or parallel upload?

Reply:
for parallel downloading of chunks.

Reply:
make sure you tried it, i remembered there is a limitation in dotnet core, they only allow 2 connections to a single host by default.

Reply:
Thanks for this reminder! Increasing connection limits will be handled inside BuildDropManager.